### ISSUE TITLE
Multiboot boot broken at Go 1.15

### ISSUE THREAD

**Issue by hugelgupf** [u-root/u-root#1803]
Booting multiboot kernels is broken at Go 1.15. If you use u-root to load multiboot kernels such as ESXi, Xen, tboot, etc, we recommend you remain at Go 1.14 until we investigate the cause.

**Comment by hugelgupf** [comment]
The culprit is https://github.com/golang/go/commit/30f8074e3510da8c39f879cfdbde600eb876a79e. It's still a bit confusing why.

I stepped through the multiboot trampoline in gdb/qemu, and we *do* actually jump into the loaded kernel. But in the case of ESXi, the loaded bootloader kernel dies somewhere relatively early on.

Setting GOAMD64=normaljumps fixes the problem -- for now. Still digging into why, still likely something to do with the trampoline.

**Comment by hugelgupf** [comment]
Haha, jk, the env var has already been removed by https://go-review.googlesource.com/c/go/+/231600/.

**Comment by hugelgupf** [comment]
Hahahaha, this has to do with our hack of inserting the right kernel entry point and the pointer to the info structure. 

TL;DR: we expect `info` to immediately follow `infotext` (same for magic and entry) in [the trampoline](https://github.com/u-root/u-root/blob/7aedd1169a0e15fed318a68eb0ffde7085abf0af/pkg/boot/multiboot/internal/trampoline/trampoline_linux_amd64.s#L118). Before https://github.com/golang/go/commit/30f8074e3510da8c39f879cfdbde600eb876a79e there was no padding between them, so that worked out. After, there is padding.

Ok, let me back up a minute.

The assembly code in [the trampoline](https://github.com/u-root/u-root/blob/7aedd1169a0e15fed318a68eb0ffde7085abf0af/pkg/boot/multiboot/internal/trampoline/trampoline_linux_amd64.s) is not executed in any Go context. Roughly, our bootloader steps look like this:

1. Go program calls kexec_load with mapping of userspace segments -> kernel segments and $entrypoint.
2. Go program calls reboot(KEXEC)
3. Kernel winds down, puts expected kernel segments into their right place in memory, jumps to $entrypoint.

Multiboot kernels expect, at their entry point, that eax = 0x2BADB002 and ebx = pointer to [multiboot_info struct](https://www.gnu.org/software/grub/manual/multiboot/multiboot.html#multiboot_002eh). Multiboot kernels also expect to be in 32bit mode without paging. When the kernel jumps out of kexec, eax/ebx may contain garbage, and we are in long-mode with paging enabled. The trampoline exists to bridge this gap. kexec jumps to the trampoline (goes to 32bit mode, turns off paging, sets up eax/ebx) and trampoline jumps to actual loaded kernel entry point.

The trampoline is loaded through kexec_load just like any other kernel segment, and kexec_load is told to jump to the trampoline entry point.

Our trampoline was originally a small gcc-assembled program, which we'd load into userspace from Go program. We'd patch that program with eax/ebx values. The build process (build a C program, include in userspace somewhere) kind of sucked, so we turned it into Go assembly. We use [runtime.funcPC](https://github.com/u-root/u-root/blob/7aedd1169a0e15fed318a68eb0ffde7085abf0af/pkg/boot/multiboot/internal/trampoline/trampoline_linux_amd64.go#L60) to determine where the trampoline starts and ends in our own address space, and pass that as a segment to kexec.

Patching works a little strangely: we embed these strings into the [assembly](https://github.com/u-root/u-root/blob/7aedd1169a0e15fed318a68eb0ffde7085abf0af/pkg/boot/multiboot/internal/trampoline/trampoline_linux_amd64.s#L118) and [search the space between trampoline.start and .end to find the string, and **replace the bytes immediately following**](https://github.com/u-root/u-root/blob/7aedd1169a0e15fed318a68eb0ffde7085abf0af/pkg/boot/multiboot/internal/trampoline/trampoline_linux_amd64.go#L87).

Unfortunately, with alignedjumps, our strings are being padded. So that BS won't work anymore, not as is.

Before:

```
TEXT infotext(SB),NOSPLIT,$0
        // u-root-info-long
        BYTE $'u'; BYTE $'-'; BYTE $'r'; BYTE $'o'; BYTE $'o';
  517620:       75 2d                   jne    51764f <info+0xf>
  517622:       72 6f                   jb     517693 <entry+0x13>
  517624:       6f                      outs   dx,DWORD PTR ds:[rsi]
        BYTE $'t'; BYTE $'-'; BYTE $'i'; BYTE $'n'; BYTE $'f';
  517625:       74 2d                   je     517654 <info+0x14>
  517627:       69 6e 66 6f 2d 6c 6f    imul   ebp,DWORD PTR [rsi+0x66],0x6f6c2d6f
        BYTE $'o'; BYTE $'-'; BYTE $'l'; BYTE $'o'; BYTE $'n';
  51762e:       6e                      outs   dx,BYTE PTR ds:[rsi]
  51762f:       67 cc                   addr32 int3 

000000000050be20 <info>:
  50be20:       00 00                   add    BYTE PTR [rax],al
  50be22:       00 00                   add    BYTE PTR [rax],al  
```

After:

```
TEXT infotext(SB),NOSPLIT,$0
        // u-root-info-long
        BYTE $'u'; BYTE $'-'; BYTE $'r'; BYTE $'o'; BYTE $'o';
  517620:       75 2d                   jne    51764f <info+0xf>
  517622:       72 6f                   jb     517693 <entry+0x13>
  517624:       6f                      outs   dx,DWORD PTR ds:[rsi]
        BYTE $'t'; BYTE $'-'; BYTE $'i'; BYTE $'n'; BYTE $'f';
  517625:       74 2d                   je     517654 <info+0x14>
  517627:       69 6e 66 6f 2d 6c 6f    imul   ebp,DWORD PTR [rsi+0x66],0x6f6c2d6f
        BYTE $'o'; BYTE $'-'; BYTE $'l'; BYTE $'o'; BYTE $'n';
  51762e:       6e                      outs   dx,BYTE PTR ds:[rsi]
// Copyright 2018 the u-root Authors. All rights reserved
  51762f:       67 cc                   addr32 int3 
  517631:       cc                      int3   
  517632:       cc                      int3   
  517633:       cc                      int3   
  517634:       cc                      int3   
  517635:       cc                      int3   
  517636:       cc                      int3   
  517637:       cc                      int3   
  517638:       cc                      int3   
  517639:       cc                      int3   
  51763a:       cc                      int3   
  51763b:       cc                      int3   
  51763c:       cc                      int3   
  51763d:       cc                      int3   
  51763e:       cc                      int3   
  51763f:       cc                      int3   

0000000000517640 <info>:
  517640:       00 00                   add    BYTE PTR [rax],al
  517642:       00 00                   add    BYTE PTR [rax],al
```

**Comment by hugelgupf** [comment]
Fix 1: depending on Go version, change the alignment value [here](https://github.com/u-root/u-root/blob/7aedd1169a0e15fed318a68eb0ffde7085abf0af/pkg/boot/multiboot/internal/trampoline/trampoline_linux_amd64.go#L36) from 0x10 to 32. (<=go1.14 is 0x10, >=go1.15 is 32)

I don't think that's sustainable.

Fix 2: We're trying to locate where info,entry, and magic are in the assembled code. We can just treat them as functions and use `funcPC` to find their location. We just need to "export" them to Go code, and not have them internal to the assembly.

I'm going with fix 2.
